### ISSUE TITLE
Nasty visual artefacts on resize

### ISSUE THREAD

**Issue by NeilMacMullen** [MicrosoftEdge/WebView2Feedback#906]
**Description**

When the control is resized, it appears to render a white background (or perhaps not render at all) while waiting for HTML rendering to return.  It would be desirable if the control itself could expose a property which allowed the background to be set to a user-specified brush/color which waiting for the HTML rendering to catch up with a resize.

Note this is also a problem when loading the control.

**Version**

WebView2: 1.0.705.50 
Framework: .Net5 (WPF)
OS: Win10 

**Repro Steps**
Create an application with a dark theme and two webviews separated by a gridsplitter.  Move the gridsplitter rapidly from left to right.  Ideally the background would remain constant even if the HTML is still rendering. 

**Screenshots**
This gif shows the effect.  Here there are 3 webview controls side by side, each running an instance of the Monaco editor.
![webview2artefacts](https://user-images.githubusercontent.com/9131337/107156404-0708b280-6976-11eb-9619-e7cc41f794f4.gif)

**Additional context**

A full project which demonstrates the issue is at https://github.com/NeilMacMullen/Textrude (easily buildable in Visual Studio)

[AB#31697759](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/31697759)

**Comment by champnic** [comment]
Hey @NeilMacMullen - we recently released an API to set the default background color. Do you want to give that a try? It requires the 1.0.790-prerelease SDK package and Canary or Dev Microsoft Edge runtime. Thanks!
https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.wpf.webview2.defaultbackgroundcolor?view=webview2-dotnet-1.0.781-prerelease

**Comment by NeilMacMullen** [comment]
@champnic   Thanks Nic, looks like my timing was quite good :-)  I installed the prerelease 1.0.790-prerelease from Nuget and the Dev Edge runtime.  The "default" behaviour seems much better; the control now seems to pick up the window background color which is nice.  However when I set
` WebView.DefaultBackgroundColor = Color.BlueViolet;` in the OnLoaded event for the containing UserControl it appears to prevent the HTML being loaded at all.  I've also tried setting it directly in XAML with the same result.  Am I missing something or can it only be set later in the process?

**Comment by champnic** [comment]
In theory it should be able to be set whenever and not break the loading/navigation, so that sounds like a bug to me. I'll get a bug created on our backlog to take a look. Where are you able to set the background color later in the process and have it work? Or is it always busted to set it explicitly?

**Comment by NeilMacMullen** [comment]
Setting it later (after the HTML content has called `window.chrome.webview.postMessage` to verify it is running) actually seems to be worse.  I get this error:
```

System.NotImplementedException
  HResult=0x80004001
  Message=Unable to cast to Microsoft.Web.WebView2.Core.Raw.ICoreWebView2Controller2.
This may happen if you are using an interface not supported by the version of the WebView2 Runtime you are using.
For instance, if you are using an experimental interface from an older SDK that has been modified or removed in a newer runtime.
Or, if you are using a public interface from a newer SDK that wasn't implemented in an older runtime.
For more information about WebView2 versioning please visit the following: https://docs.microsoft.com/microsoft-edge/webview2/concepts/versioning
  Source=Microsoft.Web.WebView2.Core
  StackTrace:
   at Microsoft.Web.WebView2.Core.CoreWebView2Controller.get__nativeICoreWebView2Controller2()
   at Microsoft.Web.WebView2.Core.CoreWebView2Controller.set_DefaultBackgroundColor(Color value)
   at Microsoft.Web.WebView2.Wpf.WebView2.DefaultBackgroundColorPropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
   at System.Windows.FrameworkElement.OnPropertyChanged(DependencyPropertyChangedEventArgs e)
   at System.Windows.DependencyObject.NotifyPropertyChange(DependencyPropertyChangedEventArgs args)
   at System.Windows.DependencyObject.UpdateEffectiveValue(EntryIndex entryIndex, DependencyProperty dp, PropertyMetadata metadata, EffectiveValueEntry oldEntry, EffectiveValueEntry& newEntry, Boolean coerceWithDeferredReference, Boolean coerceWithCurrentValue, OperationType operationType)
   at System.Windows.DependencyObject.SetValueCommon(DependencyProperty dp, Object value, PropertyMetadata metadata, Boolean coerceWithDeferredReference, Boolean coerceWithCurrentValue, OperationType operationType, Boolean isInternal)
   at System.Windows.DependencyObject.SetValue(DependencyProperty dp, Object value)
   at Microsoft.Web.WebView2.Wpf.WebView2.set_DefaultBackgroundColor(Color value)
   at TextrudeInteractive.MonacoBinding.OnReady() in D:\work\Textrude\TextrudeInteractive\Monaco\MonacoBinding.cs:line 112
   at TextrudeInteractive.MonacoBinding.OnWebMessageReceived(Object sender, CoreWebView2WebMessageReceivedEventArgs e) in D:\work\Textrude\TextrudeInteractive\Monaco\MonacoBinding.cs:line 122
   at Microsoft.Web.WebView2.Core.CoreWebView2.OnWebMessageReceived(CoreWebView2WebMessageReceivedEventArgs args)
   at Microsoft.Web.WebView2.Core.CoreWebView2WebMessageReceivedEventHandler.Invoke(ICoreWebView2 source, ICoreWebView2WebMessageReceivedEventArgs args)

  This exception was originally thrown at this call stack:
    [External Code]

Inner Exception 1:
InvalidCastException: Unable to cast COM object of type 'System.__ComObject' to interface type 'Microsoft.Web.WebView2.Core.Raw.ICoreWebView2Controller2'. This operation failed because the QueryInterface call on the COM component for the interface with IID '{C979903E-D4CA-4228-92EB-47EE3FA96EAB}' failed due to the following error: No such interface supported (0x80004002 (E_NOINTERFACE)).

```

This makes it look like the wrong SDK is being used but I'm definitely referencing the correct nuget package.

**Comment by champnic** [comment]
What version of the runtime or Microsoft Edge are you using? For new APIs in the prerelease package you'll need to use Microsoft Edge Canary or Dev (or something with a version > SDK version, so anything post-790).

**Comment by johna-ms** [comment]
Whoops! Didn't realize I never posted this. Nic, I think Neil mentioned earlier he's running Edge Dev so the runtime should be fine. That said, it seems odd to me that the control is 'picking up on' the background color of the app. The default background color is white. If you have a dark theme you should be seeing similar behavior to your buggy scenario so sounds like there's something fishy going on there. The COM casting error make me think the issue is with the runtime version and sdk version mismatching but given that you're using Edge dev that seems unlikely. Still could you post the version number of Edge Dev? Should be able to find it in ... > Help > About Edge

So same question, but more context!

**Comment by RickStrahl** [comment]
@NeilMacMullen Good luck with that :-)

This has always been a problem in WPF with Windowed controls hosted there as there is some lag between the host container control resizing and then Win32 window responding. I've played around with this for quite a bit with the old Web Browser control which had the same issues.

There are a few things that help:

* Make sure you debounce your resize events. Fewer events will make this more fluid
* Use the new WPF Resizer Preview property which lets you move the divider only while dragging and only resize when you're done.
* Use a matching background color to avoid the white flash

These are workarounds obviously but improve usability quite a bit.

**Comment by NeilMacMullen** [comment]
@champnic  @johna-ms The dev version of Edge I am using is _"Version 90.0.789.1 (Official build) dev (64-bit)"_  Having said that, this is installed alongside "classic" Edge.  Do I need to do anything special to get WebView to use this runtime or will it automatically pick up the latest version?  

@RickStrahl Thanks - some excellent points.  Did you mean the `GridSplitter.ShowsPreview` property?  Turning that on helps significantly.  (There's a minor issue that the gridsplitter can't be rendered over the Webview window but I can live with that).

Version="1.0.705.50"
![old_flash](https://user-images.githubusercontent.com/9131337/108104479-44122a80-7083-11eb-8413-2731893e6eef.gif)

Version="1.0.790-prerelease" (**not** setting DefaultBackgroundColor explicitly)

![preview_no_flash](https://user-images.githubusercontent.com/9131337/108104530-54c2a080-7083-11eb-8461-cf2c70ee9a00.gif)

**Comment by johna-ms** [comment]
Could you check in the task manager to see which runtime your app is picking up? The information will be under the 'Details' tab. You'll probably have to right click the columns, choose 'show columns', and choose 'Command line' in order to see the directory info.

**Comment by NeilMacMullen** [comment]
The answer is.... "the wrong one" (88.0.705.68) - so that's one mystery solved.  It picks up this version regardless of whether I run under VS debugger or not or run the debug or release builds.  I also tried closing down all the "old" Edge windows and only leaving Edge Dev open in case it's looking for active processes   Edge dev is installed in "C:\Program Files (x86)\Microsoft\Edge Dev\Application\90.0.789.1" which seems reasonable.

**Comment by johna-ms** [comment]
Hey Neil, you can use the CoreWebView2CreationProperties to specify a browser executable folder

https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.winforms.corewebview2creationproperties?view=webview2-dotnet-1.0.705.50

Alternatively, you could set an environment variable WEBVIEW2_RELEASE_CHANNEL_PREFERENCE=1 for whatever command line environment you are launching your app from. That option favors Canary first, then Dev, then Stable. If you don't have Canary installed, it should run Dev.

**Comment by NeilMacMullen** [comment]
Thanks John - that does indeed to the trick !  FWIW, the documentation could be a little clearer (I know it's only preview! )... From experimentation, I *believe* that the DefaultBackgroundColor is just being used for the initial control load (logical given the name) so the best "recipe" for avoiding resize glitches is...

- Set the BackgroundColor of the control/window that contains WebView to the desired color
- Set DefaultBackgroundColor of WebView to that color (or initiallly size it at 0,0) 

Anyway, thanks a lot for your help - looking forward to using this on release.  :-)

**Comment by johna-ms** [comment]
Rereading the documentation it could definitely use some work. Thank you for pointing that out.

The DefaultBackgroundColor does not need to only be for the initial control load, it can be modified at runtime. I don't think I'm hearing what you are saying. Is the unclear bit that any control containing the webview control must have a background color consistent with the webview's?

**Comment by NeilMacMullen** [comment]
Sorry, I was a bit unclear myself (not helped by probably misunderstanding differences in behaviour between the '705 and the preview.  I may well be getting it wrong again but I _think_ that when resized:

- 705  immediate resizes the wpf webview control, blitted white pixels over the new extended portion then sends a resize command to the content render process before rendering the returned bits.
- 790 (preview)  changes the ordering by sending the resize command to the render process and waiting for a new content image before resizing the wpf webview.  

I may have this wrong and perhaps 790 us just much faster at rendering after resize but the net result is that it becomes  harder to observe the webview DefaultBackgroundColor; what you actually tend to see is the background of the control it's contained within.  

Here there are 3 webviews, one in each column of a grid.  Each webview is inside a usercontrol whose background is set to a checkerboard brush. The DefaultBackgroundColor of the webview is set to Aqua.   I've turned ShowsPreview on the GridSPlitter and deliberately stressed the systems by aggressively whipping the grid-splitter from side to side. *Most* of what you see is the usercontrol background .  I had misinterpreted this in my earlier comment and thought the webview was _using_ this background explicitly rather than it just not being rendered over.


![aqua](https://user-images.githubusercontent.com/9131337/108342206-70888c80-71d2-11eb-8840-79f3de9c4d7f.gif)
 
So in summary :
- DefaultBackgroundColor works as you say it does
- Other improvements in 790 make it most relevant/likely to be seen for the initial content load (great job!) 
- To minimise visual artefacts on resize , good general advice would be:
  - Debounce resize events where possible (set ShowsPreview on GridSplitters to True)
  - Set both the DefaultBackgroundColor of the WebView *and* the background of the containing control to the same value

FWIW I can think of ways the Webview control could support this a little better; e.g. reverting to the old 705 behaviour but painting "pending" pixels in the DefaultBackgroundColor or providing its own resize event throttling but I'm happy for now :-)
